garage/tf/algos/dqn.py

```diff
@@ -0,0 +1,282 @@
+                    obs = self.env.reset()
+                    if self.use_atari_wrappers:
+                        obs = np.asarray(obs)
+
+                if self.replay_buffer.n_transitions_stored >= self.min_buffer_size:  # noqa: E501
+                    self.evaluate = True
+                    qf_loss = self.optimize_policy(itr, None)
+                    episode_qf_losses.append(qf_loss)
+                    if itr % self.target_network_update_freq == 0:
+                        self.sess.run(self._qf_update_ops, feed_dict=dict())
+
+                if self.plot:
+                    self.plotter.update_plot(self.policy, self.max_path_length)
+                    if self.pause_for_plot:
+                        input("Plotting evaluation run: Press Enter to "
+                              "continue...")
+
+                if self.evaluate:
+                    mean100ep_rewards = round(
+                        np.mean(episode_rewards[-100:]), 1)
+                    mean100ep_qf_loss = np.mean(episode_qf_losses[-100:])
+                    if itr % self.print_freq == 0:
+                        logger.record_tabular('Iteration', itr)
+                        logger.record_tabular('AverageReturn',
+                                              mean100ep_rewards)
+                        logger.record_tabular('StdReturn',
+                                              np.std(episode_rewards))
+                        logger.record_tabular('QFunction/AverageQFunctionLoss',
+                                              mean100ep_qf_loss)
+                        logger.record_tabular('EpisodeLength',
+                                              np.mean(episode_length))
+
+                logger.dump_tabular(with_prefix=False)
+
+        if created_session:
+            self.sess.close()
+
+        return mean100ep_rewards
+
+    @overrides
+    def optimize_policy(self, itr, sample_data):
+        """Optimize network using experiences from replay buffer."""
+        transitions = self.replay_buffer.sample(self.buffer_batch_size)
+
+        observations = transitions["observation"]
+        rewards = transitions["reward"]
+        actions = transitions["action"]
+        next_observations = transitions["next_observation"]
+        dones = transitions["terminal"]
+
+        if len(self.env.spec.observation_space.shape) == 3:
+            observations = observations.astype(np.float32) / 255.0
+            next_observations = next_observations.astype(np.float32) / 255.0
+
+        loss, _ = self.sess.run(
+            [self._loss, self._optimize_loss],
+            feed_dict={
+                self.qf.input: observations,
+                self.action_t_ph: actions,
+                self.reward_t_ph: rewards,
+                self.done_t_ph: dones,
+                self.target_qf.input: next_observations
+            })
+
+        return loss
+
+
+def get_target_ops(variables, target_variables):
```

Comment:
If this is required for DQN, please move the same function in DDPG to a module. Perhaps garage/tf/misc?